File: 프로그래머스/알고리즘 고득점 Kit/동적계획법(Dynamic Programming)/Level 3/N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
#include <set>
#include <cmath>
using namespace std;

int solution(int N, int number) {
    if(N == number) return 1;
    
    set<int> nums[9];
    nums[1].insert(N);
    
    int tmp = 0;
    for(int i=2; i<9; i++){
        // N을 i번 이어서 만들 수 있는 수
        tmp = 0;
        for(int j=0; j<i; tmp=(tmp*10)+N, j++);
        if(tmp == number) return i;
        else nums[i].insert(tmp);
        
        // i번 사용해서 만들 수 있는 수
        for(int j=1; j<=i/2; j++){
            set<int>::iterator it_1;
            for(it_1=nums[j].begin(); it_1!=nums[j].end(); it_1++){
                set<int>::iterator it_2;
                int k = i-j;
                
                for(it_2=nums[k].begin(); it_2!=nums[k].end(); it_2++){
                    nums[i].insert((*it_1)+(*it_2));
                    nums[i].insert((*it_1)*(*it_2));
                    if(*it_1 != *it_2) nums[i].insert(abs((*it_1)-(*it_2)));
                    if(*it_1 >= *it_2) nums[i].insert((*it_1)/(*it_2));
                    else nums[i].insert((*it_2)/(*it_1));
                }
            }
        }
        
        // number 만들었으면 리턴
        set<int>::iterator it;
        for(it=nums[i].begin(); it!=nums[i].end(); it++)
            if(*it == number) return i;
    }
    
    return -1;
}

File: 프로그래머스/알고리즘 고득점 Kit/동적계획법(Dynamic Programming)/Level 3/N으로 표현/N으로 표현.py
def solution(N, number):
    if N == number: return 1

    nums = [set() for i in range(9)]
    nums[1].add(N)

    for i in range(2, 9):
        tmp = 0
        for j in range(i):
            tmp = tmp*10+N
        if tmp == number: return i
        else: nums[i].add(tmp)

        for j in range(1, i//2+1):
            for n_1 in nums[j]:
                k = i-j
                for n_2 in nums[k]:
                    nums[i].add(n_1+n_2)
                    nums[i].add(n_1*n_2)
                    if n_1 != n_2: nums[i].add(abs(n_1-n_2))
                    if n_1 >= n_2: nums[i].add(n_1//n_2)
                    else: nums[i].add(n_2//n_1)

        if number in nums[i]: return i

    return -1
